crates/core/src/subscription/module_subscription_actor.rs

```diff
@@ -206,57 +206,61 @@ impl ModuleSubscriptionActor {
+    fn broadcast_commit_event(&self, event: ModuleEvent) -> impl Future<Output = ()> + '_ {
+        let database_update = event.status.database_update().unwrap();
+
         let auth = AuthCtx::new(self.owner_identity, event.caller_identity);
 
-        let tokio_handle = tokio::runtime::Handle::current();
-        let futures = FuturesUnordered::new();
+        let tokio_handle = &tokio::runtime::Handle::current();
 
-        let iter = self.subscriptions.par_iter_mut().map_init(
-            || {
-                scopeguard::guard(self.relational_db.begin_tx(), |tx| {
-                    let ctx = ExecutionContext::incremental_update(self.relational_db.address());
-                    self.relational_db.release_tx(&ctx, tx);
-                })
-            },
-            |tx, subscription| {
-                let database_update = event.status.database_update().unwrap();
+        let tx = &*scopeguard::guard(self.relational_db.begin_tx(), |tx| {
+            let ctx = ExecutionContext::incremental_update(self.relational_db.address());
+            self.relational_db.release_tx(&ctx, tx);
+        });
+
+        let tasks = self
+            .subscriptions
+            .par_iter()
+            .filter_map(|subscription| {
                 let incr = subscription
                     .queries
-                    .eval_incr(&self.relational_db, tx, database_update, auth)
-                    .map_err(|err| {
+                    .eval_incr(&self.relational_db, tx, database_update, auth);
+                match incr {
+                    Ok(incr) if incr.tables.is_empty() => None,
+                    Ok(incr) => Some((subscription, incr)),
+                    Err(err) => {
                         // TODO: log an id for the subscription somehow as well
-                        tracing::error!(err = &err as &dyn std::error::Error, "subscription eval_incr failed")
-                    })
-                    .ok()?;
-                if incr.tables.is_empty() {
-                    return None;
+                        tracing::error!(err = &err as &dyn std::error::Error, "subscription eval_incr failed");
+                        None
+                    }
                 }
-                Some((subscription, incr))
-            },
-        );
-        let iter = iter.filter_map(|x| x);
+            })
+            .flat_map_iter(|(subscription, incr)| {
+                let message = TransactionUpdateMessage {
+                    event: &event,
+                    database_update: incr,
+                };
+                let mut message = CachedMessage::new(message);
 
-        let iter = iter.map(|(subscription, incr)| {
-            let message = TransactionUpdateMessage {
-                event: &event,
-                database_update: incr,
-            };
-            let mut message = CachedMessage::new(message);
+                subscription.subscribers().iter().cloned().map(move |subscriber| {
```

Comment:
What do you think about doing it in single filter_map? So we don't have to clone. May be,  something like this?
```
            .filter_map(|subscription| {
                let incr = subscription
                    .queries
                    .eval_incr(&self.relational_db, tx, database_update, auth);
                match incr {
                    Ok(incr) if incr.tables.is_empty() => None,
                    Ok(incr) => {
                        let message = TransactionUpdateMessage {
                            event: &event,
                            database_update: incr,
                        };
                        let mut message = CachedMessage::new(message);
        
                        let tasks = subscription.subscribers().iter().map(move |subscriber| {
                            let message = message.serialize(subscriber.protocol);
                            tokio_handle.spawn(async move {
                                let _ = subscriber.send(message).await;
                            })
                        }).collect::<Vec<_>>();
                        Some(tasks)
                    },
                    Err(err) => {
                        // TODO: log an id for the subscription somehow as well
                        tracing::error!(err = &err as &dyn std::error::Error, "subscription eval_incr failed");
                        None
                    }
                }
            })
            .collect::<Vec<_>>();
            
```

Reply:
The spawning is cheap, it's just an Arc inc ref. And the cloning is necessary because the future passed to spawn() needs to be `'static`

Reply:
And I switched from `FuturesUnordered` to task spawning because otherwise the messages don't actually get sent until after *all* the queries are finished evaluating.

Reply:
makes sense, I agree it doesn't seems we can avoid clone but that doesn't looks like Arc inc to me.